flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java

```diff
@@ -0,0 +1,257 @@
+                            SqsExceptionClassifiers.getResourceNotFoundExceptionClassifier(),
+                            SqsExceptionClassifiers.getAccessDeniedExceptionClassifier(),
+                            SqsExceptionClassifiers.getNotAuthorizedExceptionClassifier(),
+                            getSdkClientMisconfiguredExceptionClassifier()));
+
+    private final Counter numRecordsOutErrorsCounter;
+
+    /* Url of SQS */
+    private final String sqsUrl;
+
+    /* The sink writer metric group */
+    private final SinkWriterMetricGroup metrics;
+
+    /* The asynchronous http client */
+    private SdkAsyncHttpClient httpClient;
+
+    /* The asynchronous SQS client */
+    private SqsAsyncClient sqsAsyncClient;
+
+    /* Flag to whether fatally fail any time we encounter an exception when persisting records */
+    private final boolean failOnError;
+
+    SqsSinkWriter(
+            ElementConverter<InputT, SendMessageBatchRequestEntry> elementConverter,
+            Sink.InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes,
+            boolean failOnError,
+            String sqsUrl,
+            Properties sqsClientProperties,
+            Collection<BufferedRequestState<SendMessageBatchRequestEntry>> initialStates) {
+        super(
+                elementConverter,
+                context,
+                AsyncSinkWriterConfiguration.builder()
+                        .setMaxBatchSize(maxBatchSize)
+                        .setMaxBatchSizeInBytes(maxBatchSizeInBytes)
+                        .setMaxInFlightRequests(maxInFlightRequests)
+                        .setMaxBufferedRequests(maxBufferedRequests)
+                        .setMaxTimeInBufferMS(maxTimeInBufferMS)
+                        .setMaxRecordSizeInBytes(maxRecordSizeInBytes)
+                        .build(),
+                initialStates);
+        this.failOnError = failOnError;
+        this.sqsUrl = sqsUrl;
+        this.metrics = context.metricGroup();
+        this.numRecordsOutErrorsCounter = metrics.getNumRecordsOutErrorsCounter();
+        this.httpClient = createHttpClient(sqsClientProperties);
+        this.sqsAsyncClient = createSqsClient(sqsClientProperties, httpClient);
+    }
+
+    @Override
+    protected void submitRequestEntries(
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        final SendMessageBatchRequest batchRequest =
+                SendMessageBatchRequest.builder().entries(requestEntries).queueUrl(sqsUrl).build();
+
+        CompletableFuture<SendMessageBatchResponse> future =
+                sqsAsyncClient.sendMessageBatch(batchRequest);
+
+        future.whenComplete(
+                (response, err) -> {
```

Comment:
The SQS documentation for `sendMessageBatch` is a little bit worrying...
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessageBatch.html
> The result of sending each message is reported individually in the response. Because the batch request can result in a combination of successful and unsuccessful actions, you should check for batch errors even when the call returns an HTTP status code of 200.

Have we verified that if some messages in batch fails, we will get an `err` here?

Reply:
yes, we do validated that and retry failed ones under ```handlePartiallyFailedRequest```